test/LifToken.js

```diff
@@ -132,6 +132,83 @@ contract('LifToken', function(accounts) {
     await help.checkToken(token, accounts, 125, [40,30,20,10,0]);
   });
 
+  it('should fail when calling a function that fails inside approveData', async function() {
+    let message = await Message.new();
+    help.abiDecoder.addABI(Message._json.abi);
+
+    let data = message.contract.fail.getData();
+    let transaction;
+
+    try {
+      transaction = await token.approveData(
+        message.contract.address, help.lif2LifWei(10), data,
+        {from: accounts[1]}
+      );
+    } catch (error) {
+      if (!help.isInvalidOpcodeEx(error)) throw error;
+
+      let decodedEvents = help.abiDecoder.decodeLogs(transaction.receipt.logs);
```

Comment:
probably this code is not executing: `approveData` is not supposed to throw now, it's supposed to return false. If it were throwing then `transaction` would not have been assigned and this would fail because transaction being undefined in `transaction.receipt.logs`, right?

So I think instead of the try/catch we should just assert that the return value from `approveData` equals false

Reply:
Hm yes, right now for tetsing in throwing inside the fail() function called on the approve, but it should return false instead of throw.
We cant access the return value when sending transaction, that works only for constant functions when are called and not executed via transactions.
Im going to return false on the fail() function and test that there is not events on the receipt.logs.

Reply:
there's nothing wrong in throwing inside of `fail`, that's what we need: `call` returns false if the function has thrown, true otherwise

Reply:
I see it's difficult to check whether it returned false or true... we would need another contract to issue the `approveData` and in that contract we could have some logic that depending on that return value would update some state in that contract that we can check later in the test.

In the end these `approveData`, `transferData`, etc are intended to be called from other contracts I think so it makes sense to test it that way, or am I missing something?

Reply:
So its ok, we know if it returned false or true depending on the events triggered on the transaction, we have tests where the events are triggered and we have tests where they are not, and that its because false was returned form the call. On the other commit I added the tests for that,

Reply:
On this commit, sorry, the one with the tests.

Reply:
ok, makes sense, but the try/catch is misleading, we should remove it, that code never executes. If we think the catch executes, then we should add a throw in the try block after the `approveData` call

Reply:
so there's one more piece missing here: as the catch is not running, the `assert.equal(0, decodedEvents.length);` is not running either, we have to keep that code but without the try/catch, makes sense?